accord-core/src/main/java/accord/local/ImmutableState.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package accord.local;
+
+/**
+ * Since metadata instances are immutable, there can be several copies floating around in memory, all but one
+ * of which is stale. This can lead to difficult to diagnose bugs, especially when an implementation side cache
+ * is involved.
+ * The ImmutableState base class defines the various lifecycle stage and helps restrict what the user does with it,
+ * with the goal or throwing an exception as soon as an illegal usage occurs, instead of after.
+ *
+ * An up to date instances can have one of 2 statuses, DORMANT or ACTIVE. An instance is only ACTIVE in the time
+ * between when the safe store for it's current operation is created, and closed, and is otherwise marked DORMANT.
+ * While DORMANT, it cannot be read from or updated. This is to help detect leaked instances being read from or
+ * updated when they shouldn't be.
+ *
+ * Once an update has created a superseding instance, the original instance is marked SUPERSEDED. It can then be
+ * marked CLEANING_UP, during which time its fields can be read to assist in determining what has been updated.
+ * From CLEANING_UP, an instance can be marked INVALIDATED to indicate that any access is illegal. An instance
+ * marked SUPERSEDED can also go directly to INVALIDATED, in the case of intermediate updates.
+ */
+public abstract class ImmutableState
```

Comment:
I just suggested something similar in a response to your main comment. The only scenario where I think a stale command would be useful is in the construction of an update

Reply:
> The only scenario where I think a stale command would be useful is in the construction of an update

A stale command _could_ also be useful in listeners when deciding what action to take because of a delta, though I think today we probably don't require it. We will probably need to separately track/expose the state machine types, though, as they can change within the same reference. Perhaps we can just call them an `UnsafeCommand` to distinguish them?

>What about making Command more of a state container? Something like AtomicReference, where there is no more than one instance in memory for a given Command/CommandStore

I think this would be an immediate improvement for sure. I think the only difference between our proposals is that I think it would be better to _hide_ this reference (let's call it an `UnsafeCommandRef`), as exposing it directly potentially masks some bugs. Since we expect there to be only one active operation touching the latest value at once, if the `SafeCommand` we pass around internally references an `UnsafeCommandRef` and _also_ the `UnsafeCommand` that was pointed to _when the `SafeCommand` was constructed_, then on each access we can validate that nobody has changed the underlying value, and fail an assertion if they have. Another option is to have CAS semantics, but this only catches erroneous updates and potentially makes our control flow a bit more complicated.

Not committing 100% to my proposed nomenclature, feels like we can maybe do better.

Reply:
I think probably it will be easier to discuss tweaks like this once we have the code structured this way. 

Also, let me know when you've had a chance to look at my suggestion branch and if you want to a higher bandwidth discussion. I think it may simplify things a little but it certainly affects this stuff. 

We can potentially avoid buffering state in `AbstractSafeCommandStore` since we produce a new `Command` at the end of each transformation, and push updates either through `SafeCommand` or something similar and have them take effect globally immediately, avoiding multiple copies there.

Reply:
Added LiveCommand, and it's definitely an improvement in readability and complexity. I agree that the naming isn’t my favorite, but I don’t have any better ideas at the moment. I think I may prefer SafeCommand to LiveCommand just for consistency with SafeCommandStore. Maybe CommandReference or CommandRef?

Although there isn’t an overarching builder for command updates, I have kept CommonAttributes separate which can be converted to mutable command attributes and updated via the boilerplate methods for setting things like route, partialTxn, etc, and is only converted to mutable if something needs changing. This seemed like a reasonable approach given the number of attributes set and some of the shared data, or in the case of durability, the frequency of updates. If anyone has strong feelings about that though, I can work on untangling them into some methods that return values which are passed into the command update.

Reply:
Awesome, I'll try to take a proper look over the weekend, or else Monday.

> converted to mutable command attributes

I think this is a huge improvement, though. I think we should shrink the boilerplate to {{.mutable()}}. 

I think we should also still consider removing the mutable step entirely - at least in the common path. This approach requires two copies of the sets we are updating, for instance, which is a cost we can avoid with a direct factory. It also probably makes the optimiser work harder, as each time we maybe make mutable it has to compile multiple branches that might involve copying a bunch of fields, making our methods larger and maybe impacting things like inlining.

I think, surprisingly, there aren't actually usually many fields updated and the ones we update are quite consistent. Basically it mostly happens in `set`, so we want to update `route`, `progressKey` and then whichever fields you actually need for the new state (some of deps, txn, executeAt usually). `updateHomeKey` is basically a no-op for most paths. 

I think it could make sense to keep the mutable approach for the less common updates (e.g. updating home key or progress key, which usually won't be set by themselves), and have dedicated methods for the more common ones.

If you take a look at my suggestions patch, `Commands.update` shows the main scope of the change. It can probably be cleaned up further based on your recent improvements. If you're amenable to it, I can try to rebase on what you've done here and clean it up.

Reply:
To be fair we might get most of the benefit by just having set accept a Mutable and having a dedicated add/remove listener method

Reply:
On third consideration maybe it’s better to leave further changes here until we have the commit log stuff in, as we may need to tweak things a little more then and no point overthinking this beforehand. I may see which of the more minor changes in my suggestion patch are worth retaining.

Reply:
On third consideration I think it might be better not to overthink this before some later work comes along. But, I think there remain some small clarity improvements that are possible. For instance, I don't think we necessarily need `pendingSeekableRegistrations` or `pendingSeekablesRegistrations` particularly if we move to invoking the transition method inside of `set`, so we know we have fully assembled everything, as `set` is responsible for invoking register.

Another option is to just remove `register` entirely and leave it to the implementation to manage this when we `update` the `LiveCommand`, when it already has access to a mutable set of listeners. This might be both cleaner and easier.

Reply:
Hmmm, actually we could probably move `register` to an abstract method of `LiveCommand` if we want to retain it for implementations to provide, and that would let us remove this I think.

Reply:
I like the idea of a request scoped command container, though I don't think we can restrict them to a single update per request without some reworking of `Commands#maybeExecute` as well